### ISSUE TITLE
Must attribute arguments be exactly as shown in Annex C?

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/csharpstandard#870]
**Describe the bug**

From §5 and §C.1 in the C# 7.x draft, it is not clear whether a conforming implementation has to decorate the standard classes with attributes exactly as specified in Annex C, or whether it can use different values as attribute arguments — for example, adding more AttributeTargets bits to an AttributeUsageAttribute on an attribute class.

**Example**

§C.1 specifies this type:

```csharp
[AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
    public sealed class CallerFilePathAttribute : Attribute
    {
        public CallerFilePathAttribute() { }
    }
```

Could a conforming implementation use e.g. `AttributeTargets.Parameter | AttributeTargets.Field` instead?  If so, would that count as an extension and have to be documented?  (Independently of whether the attribute has any compile-time effect when applied to a field.)

**Expected behavior**

Directly specify whether this is allowed.

**Additional context**

§5 does not mention attributes here:

> A conforming implementation of C# is permitted to provide additional types, values, objects, properties, and methods beyond those described in this specification, provided they do not alter the behavior of any strictly conforming program.

(If that paragraph were modified to allow changing attribute arguments, the "provided that" condition at the end would not be a problem, as MemberInfo.GetCustomAttributes and Attribute.GetCustomAttributes methods are not listed in Annex C and thus cannot be called by a strictly conforming program.)

§C.1 says that additional attributes are allowed, but does not mention modifications of specified attributes.  Also, that text is informative.

Inspired by <https://github.com/dotnet/csharpstandard/issues/858>; if C# 10 specifies AttributeTargets.Method on AsyncMethodBuilderAttribute but C# 7.3 does not, and modifications of attribute arguments are not allowed, then it won't be possible for implementations using the same library to comply with both versions of C#, which would be unfortunate.

**Comment by KalleOlaviNiemitalo** [comment]
This text in C.1 even looks too permissive:

> Adding attributes (other than the `ConditionalAttribute`) to existing types and members.

as adding `ObsoleteAttribute` to all types, especially with `error: true`, would make the implementation pretty useless and should not be allowed.  The text *is* informative, though.

**Comment by jskeet** [comment]
Thanks for raising this - I don't think it's blocking for C# 7 (in that there's nothing *new* about the ambiguity here), but we'll discuss it once we've shipped C# 7. If you think this really *must* be addressed before we release C# 7, please holler :)

**Comment by jskeet** [comment]
Not moving this from the C# 8 milestone as we'd still like to *try* to get it done for C# 8 - but we won't let it block the release if we're facing a deadline.

**Comment by jskeet** [comment]
(Self-assigned for the sake of triage/investigation, not actually fixing the issue...)

**Comment by jskeet** [comment]
I think this is worth a brief discussion at the next meeting where we have time to get to it.

I think we could go either way in terms of conformance, but we should be clear.

**Comment by jskeet** [comment]
Agreed in the meeting: let's get it done. (Just the bit about extra targets.)

**Comment by jskeet** [comment]
Let's tidy up the whole paragraph starting "A conforming implementation of C# is permitted to provide additional types" at the same time.
